File: desktop/src/shared/core/liveInput.ts
// Live typing straight into the pane, keystroke by keystroke — the pure core,
// ported from the Android app's ui/LiveInput.kt plus the drainer semantics of
// HuginnViewModel.sendLive and the /keys validation in huginn-appd (the wire
// contract). Deliberately NOT ported: the zero-width-sentinel IME diff. That
// exists because Android has no key events for soft keyboards, only text-field
// edits; the desktop gets real key events, so its input layer feeds ops into
// this queue directly.
//
// Why an ordered op queue at all: the path this replaced launched an
// independent request per keystroke, and independent requests are not ordered —
// type "ls" fast enough and the pane could receive "sl". A single drainer
// sending merged ops sequentially makes ordering a property of the design
// instead of a property of network luck. Fewer round trips is the bonus: a
// burst of six characters becomes one request.

/** One thing to deliver to the pane, in order. */
export type Op =
  | { readonly kind: 'text'; readonly text: string }
  | { readonly kind: 'key'; readonly keys: readonly string[] }

export const opText = (text: string): Op => ({ kind: 'text', text })
export const opKeys = (...keys: string[]): Op => ({ kind: 'key', keys })

/**
 * How long the drainer waits before draining, so a keystroke burst accumulates
 * and merges: keystrokes arrive faster than round trips complete, and merging
 * them is the point. (From HuginnViewModel.sendLive's `delay(15)`.)
 */
export const BURST_MERGE_MS = 15

/** Coalesces queued keystrokes into the fewest ops that preserve order. */
export const merge = (ops: readonly Op[]): Op[] => {
  const out: Op[] = []
  for (const op of ops) {
    const last = out[out.length - 1]
    if (op.kind === 'text' && last?.kind === 'text') {
      out[out.length - 1] = { kind: 'text', text: last.text + op.text }
    } else if (op.kind === 'key' && last?.kind === 'key') {
      out[out.length - 1] = { kind: 'key', keys: [...last.keys, ...op.keys] }
    } else {
      out.push(op)
    }
  }
  return out
}

// ----------------------------------------------------- the /keys wire contract
//
// POST /v1/sessions/<name>/keys accepts {text?, keys?[]}. The daemon REJECTS
// (400s the whole request, allowed keys included) anything outside these
// limits, so the client must never construct a request that crosses them.

/**
 * Named keys the daemon will pass to tmux. Everything else must match
 * C-<a-z>, M-<a-z>, or F1-F12; anything not matching is rejected, not passed
 * through. Mirrors huginn-appd's NAMED_KEYS + validKey().
 */
export const NAMED_KEYS: ReadonlySet<string> = new Set([
  'Enter', 'Escape', 'Tab', 'BTab', 'Space', 'BSpace', 'DC',
  'Up', 'Down', 'Left', 'Right', 'Home', 'End', 'PPage', 'NPage',
])

export const isAllowedKey = (k: string): boolean =>
  NAMED_KEYS.has(k) || /^C-[a-z]$/.test(k) || /^M-[a-z]$/.test(k) || /^F([1-9]|1[0-2])$/.test(k)

/** The daemon 400s a request with more named keys than this. */
export const MAX_KEYS_PER_REQUEST = 32

/** The daemon 400s a request with a longer text than this (UTF-16 units). */
export const MAX_TEXT_PER_REQUEST = 8000

/** One POST /keys body: exactly one of text/keys, matching how the app sends. */
export type KeysBody = { text: string } | { keys: string[] }

/**
 * Queue → wire: merge the batch, then map each op to the fewest request bodies
 * that fit the daemon's limits, preserving order throughout.
 *
 * Keys that fail the allowlist are dropped here rather than sent: one bad key
 * would 400 the entire request and take the good keys down with it.
 */
export const toWire = (ops: readonly Op[]): KeysBody[] => {
  const out: KeysBody[] = []
  for (const op of merge(ops)) {
    if (op.kind === 'text') {
      let t = op.text
      while (t.length > 0) {
        let n = Math.min(t.length, MAX_TEXT_PER_REQUEST)
        // Never split a surrogate pair across requests: the halves would land
        // as two lone surrogates and the glyph would arrive mangled.
        if (n < t.length) {
          const c = t.charCodeAt(n - 1)
          if (c >= 0xd800 && c <= 0xdbff) n--
        }
        out.push({ text: t.slice(0, n) })
        t = t.slice(n)
      }
    } else {
      const keys = op.keys.filter(isAllowedKey)
      for (let i = 0; i < keys.length; i += MAX_KEYS_PER_REQUEST) {
        out.push({ keys: keys.slice(i, i + MAX_KEYS_PER_REQUEST) })
      }
    }
  }
  return out
}

File: mobile/core/src/commonMain/kotlin/com/silencelen/huginn/ui/LiveInput.kt
package com.silencelen.huginn.ui

/**
 * Typing straight into the pane, keystroke by keystroke, instead of composing in
 * a bubble and sending.
 *
 * There is no PTY on the phone and no InputConnection to a terminal, so the trick
 * is a hidden text field holding a single invisible sentinel character. Whatever
 * the keyboard does to that field is read as a diff against the sentinel and
 * translated into keystrokes for tmux, then the field snaps back to the sentinel.
 * The sentinel is the load-bearing part: with a truly empty field most IMEs
 * express backspace as nothing at all (there is no text to delete), so deleting
 * the sentinel is how backspace becomes visible here.
 */
object LiveInput {

    /** Zero-width space: present, deletable, and invisible in the field. */
    const val SENTINEL = "​"

    /**
     * What one field change means in keystrokes.
     *
     * @param insert the text to deliver, newlines INCLUDED. The daemon sends text
     *   by bracketed paste, so an interior newline lands in a Claude composer (or
     *   a shell line) as a newline and submits nothing — which is what collapsing
     *   them was trying to achieve and did not.
     * @param enter whether a Return belongs to this edit. At most one: a tail
     *   that both begins and ends with a newline still presses Return once, since
     *   delivering fewer Returns than the input implied cannot fire one at a modal
     *   and delivering more can.
     * @param enterFirst whether that Return goes BEFORE the text. An IME that
     *   commits "\nls" pressed Return and then typed; sending it the other way
     *   round submits whatever draft the pane was holding with `ls` stuck on the
     *   end. ⚠ A caller that reads only [enter] gets the old order, which is why
     *   [ops] exists.
     */
    data class Typed(
        val backspaces: Int,
        val insert: String,
        val enter: Boolean,
        val enterFirst: Boolean = false,
    ) {
        val isNothing: Boolean get() = backspaces == 0 && insert.isEmpty() && !enter

        /** This edit as ordered ops — the whole contract, in the order it happened. */
        fun ops(): List<Op> {
            val out = mutableListOf<Op>()
            if (backspaces > 0) out += Op.Key(List(backspaces) { "BSpace" })
            if (enter && enterFirst) out += Op.Key(listOf("Enter"))
            if (insert.isNotEmpty()) out += Op.Text(insert)
            if (enter && !enterFirst) out += Op.Key(listOf("Enter"))
            return out
        }
    }

    /** One thing to deliver to the pane, in order. */
    sealed interface Op {
        data class Text(val text: String) : Op
        data class Key(val keys: List<String>) : Op
    }

    /**
     * Coalesces queued keystrokes into the fewest requests that preserve order.
     *
     * This queue exists for two reasons and the second is the important one.
     * Fewer round trips make typing feel faster — a burst of six characters
     * becomes one request. But the original path launched an independent
     * coroutine per keystroke, and independent requests are not ordered: type
     * "ls" fast enough and the pane could receive "sl". A single drainer sending
     * merged ops sequentially makes ordering a property of the design instead of
     * a property of network luck.
     */
    fun merge(ops: List<Op>): List<Op> {
        val out = ArrayList<Op>(ops.size)
        for (op in ops) {
            val last = out.lastOrNull()
            if (op is Op.Text && last is Op.Text) {
                out[out.size - 1] = Op.Text(last.text + op.text)
            } else if (op is Op.Key && last is Op.Key) {
                out[out.size - 1] = Op.Key(last.keys + op.keys)
            } else {
                out.add(op)
            }
        }
        return out
    }

    /**
     * A field that is known to have been resting at the sentinel.
     *
     * ⚠ ONLY SAFE WHEN NOTHING CAN BE IN FLIGHT — the two-argument [diff] is
     * the one a live IME talks to, and the reason is written on it. This
     * overload is the resting case: a first edit, and the tests that describe
     * one.
     */
    fun diff(newValue: String): Typed = diff(SENTINEL, newValue)

    /**
     * Diffs the field against THE TEXT WE LAST SAW IN IT, not against a constant.
     *
     * ⚠⚠ THE SENTINEL IS A RESTING VALUE, NOT A BASELINE, AND CONFUSING THE TWO
     * DUPLICATED EVERY FAST BURST. The field used to be snapped back to [SENTINEL]
     * inside `onValueChange` and every change diffed against that constant, so a
     * second IME change that arrived before the snap-back had round-tripped
     * through recomposition saw the CUMULATIVE buffer and re-emitted everything
     * typed so far. Five characters delivered in one burst arrived as fifteen —
     * 1+2+3+4+5, the triangular number that named the bug — straight into a live
     * Claude Code prompt. Diffing against the last observed text makes a
     * mid-burst change mean exactly the characters it added, whatever the field
     * still holds and whenever the reset lands.
     *
     * Pure and deliberately paranoid: an IME may rewrite the whole field (paste,
     * autocorrect, voice input), so this never assumes the change was a single
     * character. Anything [previous] had past the shared prefix was deleted, and
     * it is reported as that many backspaces — including the sentinel itself,
     * which is how a backspace at the very start of an empty field becomes
     * visible at all.
     *
     * ⚠ NEWLINES ARE STRUCTURE, NOT A RETURN KEY. This used to delete every
     * newline in the tail and press Return once at the end, so a three-line
     * clipboard arrived as `git statusgit log --onelinels -la` and was SUBMITTED:
     * a command nobody wrote, run in a live shell. Only a newline at the very
     * start or the very end of the edit is a Return now; the ones in the middle
     * stay in the text, where bracketed paste delivers them without submitting.
     * `\r\n` and a bare `\r` are normalised first — the old collapse left carriage
     * returns in `send-keys -l` text, delivering the mid-text Return it claimed
     * to prevent.
     */
    fun diff(previous: String, current: String): Typed {
        if (current == previous) return Typed(0, "", false)
        val shared = sharedPrefix(previous, current)
        val backspaces = previous.length - shared
        val tail = normalizeNewlines(current.substring(shared))

        val trailing = tail.endsWith("\n")
        val body = if (trailing) tail.dropLast(1) else tail
        val leading = body.startsWith("\n")
        val insert = if (leading) body.drop(1) else body
        return Typed(
            backspaces = backspaces,
            insert = insert,
            enter = trailing || leading,
            // Only when there is text for it to come before; a bare newline is
            // just Return, and saying it came "first" would be noise.
            enterFirst = leading,
        )
    }

    /**
     * How much of [a] and [b] is the same run of characters from the start.
     *
     * ⚠ NEVER SPLITS A SURROGATE PAIR. Half a code point is not a keystroke, and
     * an emoji typed into the pane would otherwise be reported as one backspace
     * and a lone low surrogate.
     */
    private fun sharedPrefix(a: String, b: String): Int {
        val n = minOf(a.length, b.length)
        var i = 0
        while (i < n && a[i] == b[i]) i++
        if (i in 1..(a.length - 1) && a[i - 1].isHighSurrogate()) i--
        return i
    }

    private fun normalizeNewlines(s: String): String =
        if ('\r' !in s) s else s.replace("\r\n", "\n").replace('\r', '\n')
}
